### DISCUSSION TITLE
[General] RPCs return baseFeePerGas / maxFeePerGas that is too low and lags behind actual gasPrice

### DISCUSSION THREAD

**Discussion by Antlion12** [zkSync-Community-Hub/zksync-developers#841] [General] [answered]
### Team or Project

Treasure

### Environment

Mainnet

### L2 block number

110100

## The Problem
Once every 1--2 days, our RPCs for Treasure mainnet and Treasure Topaz testnet return `maxFeePerGas` numbers (based on the historical `baseFeePerGas`) that are lower than the current `gasPrice`, which in turn causes ` transactions to fail with the error that `max fee per gas less than block base fee`.

This error doesn't occurs with other RPCs (Ethereum mainnet, Sepolia, Arbitrum, etc).

## Proposed solution we'd love to see from ML

Have the ZK-based RPCs return a `baseFeePerGas` in the `eth_getBlock` call that adapts more quickly to what is returned in the `eth_getGasPrice` RPC call.

## Steps to reproduce:

(1) Run `npx ts-node`:

(2) Enter the following commands in the REPL.
```
> import { ethers } from 'ethers';
> const feeData = await new ethers.JsonRpcProvider('https://rpc.treasure.lol').getFeeData();

> console.log(feeData);
FeeData {
  gasPrice: 1518863251238n,
  maxFeePerGas: 1458842327386n,
  maxPriorityFeePerGas: 0n
}

> feeData.gasPrice > feeData.maxFeePerGas
true

> ethers.formatEther(feeData.gasPrice);
'0.000001518863251238'

> ethers.formatEther(feeData.maxFeePerGas);
'0.000001458842327386'
```

## Root cause

The `eth_getBlock` call's `baseFeePerGas` (which is used by `ethers` to compute the `maxFeePerGas`) seems to lag behind the gas price returned by `eth_gasGasPrice` (used by `ethers` to compute the `gasPrice`).

## Current workaround (suboptimal, poor DX)

We've been telling partners to manually set `maxFeePerGas = 2 * gasPrice` for their `send()` calls, but this is not ideal because they never had to do this before with other RPC providers. However, this is not good for DX because it requires partners to refactor all their RPC calls to manually set the `maxFeePerGas`.

**Comment by alhenaai** [comment]
It seems you are experiencing issues with the `baseFeePerGas` and `maxFeePerGas` values returned by the RPCs for the Treasure mainnet and Treasure Topaz testnet, which are not adapting quickly enough to the current gas prices. This discrepancy is causing transaction failures due to the `max fee per gas less than block base fee` error.

For ZKsync, the handling of `baseFeePerGas` and `maxFeePerGas` is indeed different from Ethereum's typical approach. In ZKsync, while the EIP-1559 transaction format is supported, the `maxFeePerGas` and `maxPriorityFeePerGas` parameters are not utilized in the same way. This means that transactions might not behave as expected if they rely on dynamic fee adjustments similar to Ethereum's mainnet.

Given the specific nature of your issue with the Treasure networks, which might not directly relate to ZKsync's typical operations, I recommend reaching out to the support or development team responsible for the Treasure mainnet and Treasure Topaz testnet. They might be able to provide a more tailored solution or update to address the lag in `baseFeePerGas` updates.

For more detailed discussions or to explore potential solutions with other developers, you might find it helpful to visit the [ZKsync developers discussions on Github](https://github.com/zkSync-Community-Hub/zksync-developers/discussions).

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by JackHamer09** [comment]
Thanks for the report! Will investigate and get back to you asap!

**Comment by shahar4** [comment, marked as answer]
@Antlion12 thanks for this wonderfully detailed report!

This is a known issue that's being worked on ([PR](https://github.com/matter-labs/zksync-era/pull/2868)). For added context:
- This scenario is somewhat of an edge case, occurring only at the start of a new L1 batch if there's a significant time gap between the last block of the previous batch and the first transaction of the new one such that there can be more than 2x gas price discrepancies. Importantly, this means that this is not an issue once there's more traffic.
- It's also a bit non trivial to fix, so it's taking a bit longer.

**Reply by itegulov** [reply]
Small correction to what Shahar said above: the PR will not help with this specific problem as `ethers` will not utilize open batch's pricing. The two problems are related but a fix for this would have to come in separately once we align internally on the best way to make sure two consecutive block's gas price is never more than 2x off.
